### ISSUE TITLE
Unable to link a framework as an export dependency to a static library on MacOs

### ISSUE THREAD

**Issue by AnandR-2D2** [build2/build2#150]
I used a configuration to build static libraries with a project that exports a framework as a dependency.
`c.exports.libs += -framework OpenGL`.

However, linker command line does not contain the required arguments when linking the static library against the test.
When building shared libs, `-framework OpenGL` is passed properly to the linker.

Example repository (attempt to package `glew`): https://github.com/AnandR-2D2/glew.git
Configuration to build only the static library:
`bdep init -C @clang-static cc config.c=clang config.bin.lib=static` ----> Fails to link the test command

Configuration to build only the shared library:
`bdep init -C @clang-shared cc config.c=clang config.bin.lib=shared` ----> Has the right options passed

**Comment by boris-kolpackov** [comment]
The `*.exports.libs` variables should only contain interface dependencies while `*.libs` should contain both interface and implementation dependencies. The former is used when a client is linking a shared library (since the implementation part is already linked to the shared library) while the latter is used when linking a static library (since in this case we need everything). So you need to decide whether the OpenGL library/framework is an interface or implementation dependency of your library (see the build system manual for details on the distinction). Then change:

```
if ($c.target.class == 'macos')
{
  lib{glew}: c.export.libs += -framework OpenGL
}
else
{
  lib{glew}: c.export.libs += -lopengl
}
```

To:

```
if ($c.target.class == 'macos')
{
  lib{glew}: c.export.libs += -framework OpenGL
  c.libs += -framework OpenGL
}
else
{
  lib{glew}: c.export.libs += -lopengl
  c.libs += -lopengl
}
```

If it's an interface dependency. And to:

```
if ($c.target.class == 'macos')
{
  c.libs += -framework OpenGL
}
else
{
  c.libs += -lopengl
}
```

If it's an implementation dependency.

**Comment by AnandR-2D2** [comment]
Hi @boris-kolpackov, I would like to confirm your solution works!!
However, I have misunderstood some concepts.
In reality, `glew` does not reference any functions that require `OpenGL`. As such, `OpenGL` is not needed to build `lib{glew}` either as a shared or static library. However, `exe{glewinfo}` (a test) references `OpenGL` functions and therefore requires to be linked against it.

`OpenGL` should therefore be added as a dependency to `exe{glewinfo}` and removed from `lib{glew}`'s dependency list.

However, it's a good point that you brought up that `c.export.libs` does not act on static libraries.
Is the understanding then correct that using `lib{glew}` in this context pass the option only to `libs{glew}` and the option is ignored for `liba{glew}`?

**Comment by boris-kolpackov** [comment]
> Is the understanding then correct that using lib{glew} in this context pass the option only to libs{glew} and the option is ignored for liba{glew}?

Not quite. In order to build the library itself you always need to link it with both interface and implementation dependencies (while you can probably construct a degenerate case where a library does not need to be linked to its interface dependencies, in `build2` we ignore such cases). You specify such dependencies either as prerequisites or in the `*.libs` variable.

When you link the shared variant of this library to an executable, then the executable need to also explicitly link interface dependencies. This is what `*.export.libs` are for.

Finally, when you link the static variant of this library to an executable, then both interface and implementation dependencies must be linked, and they are taken from the same place as when building the library (prerequisites and `*.libs`). Note that this is just how static linking happened to work (`.a`/`.lib` are dumb archives of object files and do not include any information about other libraries they may depend on). In another universe this could have been implemented differently.
